tests/version/tests.py

```diff
@@ -12,6 +12,18 @@ def test_development(self):
         ver_string = get_version(ver_tuple)
         self.assertRegex(ver_string, r'1\.4(\.dev[0-9]+)?')
 
+    def test_development_no_file(self):
+        import django.utils.version
+        if not hasattr(django.utils.version, '__file__'):
+            return  # test_development() already accomplished this test's goal
```

Comment:
As we run the test suite, does this ever evaluate True? (i.e. Do we need this branch?)

Reply:
I think [eventually](https://code.djangoproject.com/ticket/32316#comment:4) I may want to [propose](https://code.djangoproject.com/ticket/32316#comment:6) that Django run tests (or at least _be able to be tested_) in a frozen environment. I agree that for now this branch won't get used. But it seems to me to that it is of at most de minimis cost to include this branch for technical correctness—the whole point of this patch is that [`__file__`](https://docs.python.org/3/reference/import.html#__file__) may not exist—to support testing in a frozen environment in the future.

That said, I won't die on this hill. If you want the branch gone, I'll take it out.

Reply:
Use `skipIf()`.

Reply:
I added `skipIf()` in 17c18e126cc8572b20424268d59c73223e299263.

But then I realized that what I cared about here was the test passing when `__file__` is missing, not avoiding duplication with `test_development()`. 40059b2b988fba0417a3c768eb87e29850f19504 simplifies this a little (it's 3 lines shorter) by allowing `test_development_no_file` to run when `__file__` is missing.

It [doesn't matter yet](https://github.com/django/django/pull/13841#discussion_r570230164), but I manually tested that both commits work in the absence of `__file__` with the following very stupid patch:

```diff
diff --git a/tests/runtests.py b/tests/runtests.py
index afdf68c21b..6e13ec3ad5 100755
--- a/tests/runtests.py
+++ b/tests/runtests.py
@@ -29,6 +29,9 @@ else:
     )
     from django.utils.log import DEFAULT_LOGGING
 
+import django.utils.version
+del django.utils.version.__file__
+
 try:
     import MySQLdb
 except ImportError:
```

and running `./runtests.py version` from the `tests` directory.

Reply:
> But then I realized that what I cared about here was the test passing when `__file__` is missing, not avoiding duplication with `test_development()`.

`test_development()` already checks this scenario. I don't see much value in duplicated tests.